draft-pda-core.md

```diff
@@ -291,11 +291,437 @@ Each PA task in this document is divided into three sub-protocols as follows.
+
+After the client uploads a report to the leader, the leader and helper verify
+the validity proof as specified by the PA protocol. This phase involves two
+requests sent from by the leader to the helper: the *verify start request*, and
+the *verify finish request*. The contents of each request depend largely on the
+specific PA protocol. However, both requests contain the helper's share of the
+report, allowing the helper to run the verification protocol statelessly.
+
+Another important feature of the verification protocol is that the leader may
+"batch" multiple runs of the protocol into the same message flow. This allows
+the leader to throttle the load on the helper endpoint.
+
+At the end of this phase, the leader and helper will have decided whether a set
+of client inputs are valid. For each valid input, they proceed as described in
+{{pa-collect}}.
+
+### Verify Start
+
+The leader begins the protocol by collecting a sequence of helper shares that
+all correspond to the same helper endpoint URL, key configuration, and PA task.
+For each share, it generates the first protocol messages as specified by the PA
+protocol, then proceeds as follows.
+
+**Request.**
+Let `[helper]` denote the helper URL. The leader sends a POST request to
+`[helper]/verify_start` with the following payload:
+
+~~~
+struct {
+  PATask task;
+  uint8 key_config_id;
+  PAHelperShare shares<0..2^24-1>;       // Any number of helper shares
+  select (PAVerifyStartReq.task.proto) { // Payload for each helper share
+    case prio: PrioVerifyStartReq payloads<0..2^24-1>;
+    case hits: HitsVerifyStartReq payloads<0..2^24-1>;
+  }
+} PAVerifyStartReq;
+~~~
+
+Each helper share has a corresponding message payload contained in
+`PAVerifyStartReq.payloads`. This includes any protocol-specific information the
+helper needs for the first step of input validation, e.g., the joint randomness
+used by the leader and helper for the protocol run. Note that a well-formed
+message contains as many payloads as shares, i.e., the sequence
+`PAVerifyStartReq.payloads` has the same number of elements as
+`PAVerifyStartReq.shares`. Moreover, the `i`-th payload should correspond to the
+`i`-th share for each `0 < i <= L`, where `L` is the length of
+`PAVerifyStartReq.payloads`.
+
+[TODO: Instead of a sequence of shares and a sequence of payloads,
+`PAVerifyStartReq` should have a sequence of (share, payload) pairs. (Here and
+below.) This is difficult to express in TLS syntax because we have to select on
+the protocol type.]
+
+**Response.**
+The helper handles well-formed requests as follows. (As usual, malformed
+requests are handled as described in {{pa-error-common-aborts}}.) It first looks
+for the PA parameters `PAParam` for which `PAVerifyStartReq.task.id ==
+PAParam.task.id`. Next, it looks up the HPKE config and corresponding secret key
+associated with `PAVerifyStartReq.key_config_id`. If not found, then it aborts
+and alerts the leader with "unrecognized key config".
```

Comment:
What does the leader do for a batch in this situation?

Reply:
Ah, good point. This might happen, say, if the helper advertises the wrong key config during the upload phase. The shares are useless if the helper can't decrypt them, so I don't think the leader has a choice other than to abort. This falls into the bucket of situations that re-running the protocol with multiple helpers is meant to address.

Reply:
Yeah, we might want to note this (or flag a TODO to spell it out later).